### PR TITLE
[FLINK-15959][flink-runtime] Add min number of slots configuration to limit total number of slots

### PR DESCRIPTION
**Pull request by xiangyuf**
## What is the purpose of the change

Add min number of slots configuration to limit total number of slots

## Brief change log

- Introduce "slotmanager.number-of-slots.min" configuration option with default value 0
- Introduce "slotmanager.min-total-resource.cpu" and "slotmanager.min-total-resource.memory" configuration option with default value derived from "slotmanager.number-of-slots.min"
- Make DefaultResourceAllocationStrategy respect the min number of cpu and memory. DefaultResourceAllocationStrategy will alway check the min resource requirement when reconciling cluster resources or fulfilling resource requirement.

## Verifying this change

This change added tests and can be verified as follows:

  - Added unit tests **SlotManagerConfigurationTest#testComputeMinTotalCpu**,  **SlotManagerConfigurationTest#testComputeMinTotalMemory**, **SlotManagerConfigurationTest#testComputeMinMaxSlotNumIsValid**, **SlotManagerConfigurationTest#testComputeMinMaxSlotNumIsInValid**, **SlotManagerConfigurationTest#testComputeMinMaxMemoryIsInValid**, **SlotManagerConfigurationTest#testComputeMinMaxCpuIsInValid** to test new added options
  - Added unit tests **DefaultResourceAllocationStrategyTest#testMinRequiredCPULimitInTryReconcile**, **DefaultResourceAllocationStrategyTest#testMinRequiredMemoryLimitInTryReconcile**, **DefaultResourceAllocationStrategyTest#testMinRequiredCPULimitInTryFulfill** and **DefaultResourceAllocationStrategyTest#testMinRequiredMemoryLimitInTryFulfill**  to test the function in DefaultResourceAllocationStrategy

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) yes
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) yes, it will affect the behavior of SlotManager in  JobManager
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) yes
  - If yes, how is the feature documented? [FLIP-362](https://cwiki.apache.org/confluence/display/FLINK/FLIP-362%3A+Support+minimum+resource+limitation)

### CODE REVIEW

**Comment by xiangyuf** [comment]
@flinkbot run azure

**Comment by xiangyuf** [comment]
@flinkbot run azure

**Comment by xiangyuf** [comment]
@flinkbot run azure

**Comment by xiangyuf** [comment]
Hi, CI has already passed. Pls view when u have time? @KarmaGYZ @huwh

**Comment by xiangyuf** [comment]
@flinkbot run azure

**Review by KarmaGYZ — CHANGES_REQUESTED** [review]
Thanks for the PR, Xiangyu! I give it a first pass and left some comments.

My major concern is that, place the resource allocation decisions related to the minimum resources in the SlotManager will increase the coupling between the SlotManager and ResourceAllocationStrategy. If we would implement a new ResourceAllocationStrategy for hetorogenous TM, the resource spec of TMs to fulfill min resouce restriction should be provided by that specific strategy.

So, I think it make more sense to handing over the decision-making authority to `ResourceAllocationStrategy`. Maybe we can rename the `tryReleaseUnusedResources` to `tryReconcileClusterResources` and extend it capability to resource allocation as well. Also, we can rename the `checkTaskManagerReleasable` to `reconcileClusterResources`.

cc @huwh

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 1]
```diff
@@ -83,6 +87,24 @@ public DefaultResourceAllocationStrategy(
...
             int redundantTaskManagerNum) {
+        this(
+                totalResourceProfile,
+                numSlotsPerWorker,
+                evenlySpreadOutSlots,
+                taskManagerTimeout,
+                redundantTaskManagerNum,
+                new CPUResource(0.0),
+                MemorySize.ZERO);
+    }
```
Comment (KarmaGYZ) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:98]: Should be `@VisibleForTesting`. BTW, I think it's overkill to add a new constructor only for test here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 2]
```diff
@@ -359,24 +425,45 @@ private void tryFulFillRedundantResources(
...
-            totalAvailableResources = totalAvailableResources.merge(totalResourceProfile);
+            resourceAvailable = resourceAvailable.merge(totalResourceProfile);
+            resourceInTotal = resourceInTotal.merge(totalResourceProfile);
         }
     }
 
+    private ResourceProfile getTotalResourceOfTaskManagers(List<TaskManagerInfo> taskManagers) {
+        return taskManagers.stream()
+                .map(TaskManagerInfo::getTotalResource)
+                .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
```
Comment (KarmaGYZ) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:457]: In this strategy, all the TM resource spec are the same. we can just use it to multiply the TM num for efficiency.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 3]
```diff
@@ -359,24 +425,45 @@ private void tryFulFillRedundantResources(
...
+    }
+
+    private ResourceProfile getTotalResourceOfPendingTaskManagers(
+            List<PendingTaskManager> pendingTaskManagers) {
+        return pendingTaskManagers.stream()
+                .map(PendingTaskManager::getTotalResourceProfile)
+                .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
```
Comment (KarmaGYZ) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:464]: ditto.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 4]
```diff
@@ -359,24 +425,45 @@ private void tryFulFillRedundantResources(
...
                         .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
 
-        tryFulFillRedundantResourcesWithAction(
-                requiredRedundantResource,
+        ResourceProfile totalResources =
+                Stream.concat(
+                                availableRegisteredResources.stream(),
+                                availablePendingResources.stream())
+                        .map(internalResourceInfo -> internalResourceInfo.totalProfile)
+                        .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
```
Comment (KarmaGYZ) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:433]: Ditto. totalProfile * TM_num.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 5]
```diff
@@ -347,8 +376,45 @@ && canFulfillRequirement(effectiveProfile, remainResource)) {
         }
     }
 
-    private void tryFulFillRedundantResources(
-            ResourceProfile requiredRedundantResource,
+    private void tryFulFillMinRequiredResources(
```
Comment (KarmaGYZ) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:379]: Seems to be not used anywhere.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java (hunk 3 above)
Comment (xiangyuf) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:464]: @KarmaGYZ thanks for reviewing. IMHO we better not making this assumption here and the efficiency here is not the top priority.

**Review by huwh — COMMENTED** [review]
Thanks for your contribution. I haven't completed the review yet.
I left some comments. Please take a look.

**Comment by xiangyuf** [comment]
> Thanks for the PR, Xiangyu! I give it a first pass and left some comments.
> 
> My major concern is that, place the resource allocation decisions related to the minimum resources in the SlotManager will increase the coupling between the SlotManager and ResourceAllocationStrategy. If we would implement a new ResourceAllocationStrategy for hetorogenous TM, the resource spec of TMs to fulfill min resouce restriction should be provided by that specific strategy.
> 
> So, I think it make more sense to handing over the decision-making authority to `ResourceAllocationStrategy`. Maybe we can rename the `tryReleaseUnusedResources` to `tryReconcileClusterResources` and extend it capability to resource allocation as well. Also, we can rename the `checkTaskManagerReleasable` to `reconcileClusterResources`.
> 
> cc @huwh

@KarmaGYZ Resolved.

**Comment by xiangyuf** [comment]
@KarmaGYZ @huwh Hi, the proposed flip has been accepted, u can continue the review when u have time.

**Review by KarmaGYZ — APPROVED** [review]
Thanks for the update, @xiangyuf . I find some typos and addressed it myself. The PR now LGTM.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 6]
```diff
@@ -347,10 +367,26 @@ && canFulfillRequirement(effectiveProfile, remainResource)) {
         }
     }
 
-    private void tryFulFillRedundantResources(
-            ResourceProfile requiredRedundantResource,
+    private boolean isRequiredResourcesFulfilled(
+            ResourceProfile availableResourceProfile, ResourceProfile totalResourceProfile) {
```
Comment (huwh) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:371]: We already have a class variable `totalResourceProfile`, It's better to not use this in method arguments.

```suggestion
            ResourceProfile resourceAvailable, ResourceProfile resourceInTotal) {
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 7]
```diff
@@ -347,10 +367,26 @@ && canFulfillRequirement(effectiveProfile, remainResource)) {
...
+        return isRedundantResourcesFulfilled(availableResourceProfile)
+                && isMinRequiredResourcesFulfilled(totalResourceProfile);
+    }
+
+    private boolean isRedundantResourcesFulfilled(ResourceProfile availableResourceProfile) {
```
Comment (huwh) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:376]: ditto

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java [hunk 8]
```diff
@@ -347,10 +367,26 @@ && canFulfillRequirement(effectiveProfile, remainResource)) {
...
+        return availableResourceProfile.allFieldsNoLessThan(
+                totalResourceProfile.multiply(redundantTaskManagerNum));
+    }
+
+    private boolean isMinRequiredResourcesFulfilled(ResourceProfile totalResourceProfile) {
```
Comment (huwh) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java:381]: ditto

**Review by huwh — COMMENTED** [review]
Thanks for the update, @xiangyuf , I left one comment, please take a look

**Comment by xiangyuf** [comment]
> Thanks for the update, @xiangyuf , I left one comment, please take a look

hi @huwh , I changed the variable name to `resourcesAvailable` and `resourcesInTotal` to represent cluster level available resources and total resources.

**Review by huwh — APPROVED** [review]
@xiangyuf thanks for your quick update. LGTM

**Pull request merged**